servers/src/main/java/tachyon/master/lineage/LineageMaster.java

```diff
@@ -0,0 +1,415 @@
+import tachyon.master.MasterBase;
+import tachyon.master.MasterContext;
+import tachyon.master.file.FileSystemMaster;
+import tachyon.master.journal.Journal;
+import tachyon.master.journal.JournalEntry;
+import tachyon.master.journal.JournalOutputStream;
+import tachyon.master.lineage.checkpoint.CheckpointPlan;
+import tachyon.master.lineage.checkpoint.CheckpointPlanningExecutor;
+import tachyon.master.lineage.journal.AsyncCompleteFileEntry;
+import tachyon.master.lineage.journal.LineageEntry;
+import tachyon.master.lineage.journal.LineageIdGeneratorEntry;
+import tachyon.master.lineage.journal.PersistFilesEntry;
+import tachyon.master.lineage.journal.RequestFilePersistenceEntry;
+import tachyon.master.lineage.meta.Lineage;
+import tachyon.master.lineage.meta.LineageFile;
+import tachyon.master.lineage.meta.LineageFileState;
+import tachyon.master.lineage.meta.LineageIdGenerator;
+import tachyon.master.lineage.meta.LineageStore;
+import tachyon.master.lineage.meta.LineageStoreView;
+import tachyon.master.lineage.recompute.RecomputeExecutor;
+import tachyon.master.lineage.recompute.RecomputePlanner;
+import tachyon.thrift.BlockInfoException;
+import tachyon.thrift.BlockLocation;
+import tachyon.thrift.CheckpointFile;
+import tachyon.thrift.CommandType;
+import tachyon.thrift.FileAlreadyExistException;
+import tachyon.thrift.FileBlockInfo;
+import tachyon.thrift.FileDoesNotExistException;
+import tachyon.thrift.InvalidPathException;
+import tachyon.thrift.LineageCommand;
+import tachyon.thrift.LineageDeletionException;
+import tachyon.thrift.LineageDoesNotExistException;
+import tachyon.thrift.LineageInfo;
+import tachyon.thrift.LineageMasterService;
+import tachyon.util.ThreadFactoryUtils;
+import tachyon.util.io.PathUtils;
+
+/**
+ * The lineage master stores the lineage metadata in Tachyon, and it contains the components that
+ * manage all lineage-related activities.
+ */
+public final class LineageMaster extends MasterBase {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private final TachyonConf mTachyonConf;
+  private final LineageStore mLineageStore;
+  private final FileSystemMaster mFileSystemMaster;
+  private final LineageIdGenerator mLineageIdGenerator;
+
+  /** The service that checkpoints lineages. */
+  private Future<?> mCheckpointExecutionService;
+  /** The service that recomputes lineages. */
+  private Future<?> mRecomputeExecutionService;
+
+  /** Map from worker to the files to checkpoint on that worker. Used by checkpoint service. */
+  private final Map<Long, Set<LineageFile>> mWorkerToCheckpointFile;
+
+  /**
+   * @param baseDirectory the base journal directory
+   * @return the journal directory for this master
+   */
+  public static String getJournalDirectory(String baseDirectory) {
+    return PathUtils.concatPath(baseDirectory, Constants.LINEAGE_MASTER_SERVICE_NAME);
+  }
+
+  public LineageMaster(Journal journal, FileSystemMaster fileSystemMaster) {
```

Comment:
NIT: All other masters' constructors take journal as the last parameter, better to keep the consistency